# Ebay Germany Used Card Dataset analysis


In [1]:
import numpy as np
import pandas as pd

# Print the pandas version and the configuration
print(pd.__version__)

1.0.1


The dataset data dictionary provided with data is as follows:

-dateCrawled - When this ad was first crawled. All field-values are taken from this date.
-name - Name of the car.
seller - Whether the seller is private or a dealer.
offerType - The type of listing
price - The price on the ad to sell the car.
abtest - Whether the listing is included in an A/B test.
vehicleType - The vehicle Type.
yearOfRegistration - The year in which which year the car was first registered.
gearbox - The transmission type.
powerPS - The power of the car in PS.
model - The car model name.
kilometer - How many kilometers the car has driven.
monthOfRegistration - The month in which which year the car was first registered.
fuelType - What type of fuel the car uses.
brand - The brand of the car.
notRepairedDamage - If the car has a damage which is not yet repaired.
dateCreated - The date on which the eBay listing was created.
nrOfPictures - The number of pictures in the ad.
postalCode - The postal code for the location of the vehicle.
lastSeenOnline - When the crawler saw this ad last online.

In [2]:
autos = pd.read_csv("data/autos.csv",encoding='Latin-1')

autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [3]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

### Clean Columns


In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

### update column names


In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

In [6]:
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price ab_test vehicle_type  registration_year    gearbox  power_ps  model  \
0    480    test          NaN               1993    manuell         0   golf   
1  18300    test        coupe               2011    manuell       190    NaN   
2   9800    test          suv               2004  automatik       163  grand   
3   1500    test   kleinwagen               2001    manuell        75   golf   
4   3600    test   kleinwagen               2008    manuell        69  fabia   

   odometer  registration_month fuel_type       brand unrepaired_damage  \
0    150000                   0    benzin  volkswagen               NaN   
1    125000                   5    diesel        audi                ja   
2    125000                   8    diesel        jeep               NaN   
3    150000                   6    benzin  volkswagen              nein   
4     90000                   7    diesel       skoda              nein   

            ad_created  num_photos  postal_code            last_seen  
0  2016-03-24 00:00:00           0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00           0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00           0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00           0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00           0        60437  2016-04-06 10:17:21

### Initial Data Exploration and Cleaning¶



In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model       odometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_photos   postal_code  \
count             299468               371528    371528.0  371528.00000   
unique                 2                  114         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq              263182                14450         NaN           NaN   
mean                 NaN                  NaN         0.0   50820.66764   
std                  NaN                  NaN         0.0   25799.08247   
min                  NaN                  NaN         0.0    1067.00000   
25%                  NaN                  NaN         0.0   30459.00000   
50%                  NaN                  NaN         0.0   49610.00000   
75%                  NaN                  NaN         0.0   71546.00000   
max                  NaN                  NaN         0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

initial observations:

There are a number of text columns where all (or nearly all) of the values are the same:
seller
offer_type
The num_photos column looks odd, we'll need to investigate this further.

In [8]:
autos['num_photos'].value_counts()

0    371528
Name: num_photos, dtype: int64

It looks like the num_photos column has 0 for every column. We'll drop this column, plus the other two we noted as mostly one value.



In [9]:
autos = autos.drop(["num_photos","offer_type","seller"],axis=1)

There are two columns, price and odometer, which are numeric values with extra characters being stored as text. We'll clean and convert these.



In [10]:
autos.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 17 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   date_crawled        371528 non-null  object
 1   name                371528 non-null  object
 2   price               371528 non-null  int64 
 3   ab_test             371528 non-null  object
 4   vehicle_type        333659 non-null  object
 5   registration_year   371528 non-null  int64 
 6   gearbox             351319 non-null  object
 7   power_ps            371528 non-null  int64 
 8   model               351044 non-null  object
 9   odometer            371528 non-null  int64 
 10  registration_month  371528 non-null  int64 
 11  fuel_type           338142 non-null  object
 12  brand               371528 non-null  object
 13  unrepaired_damage   299468 non-null  object
 14  ad_created          371528 non-null  object
 15  postal_code         371528 non-null  int64 
 16  la

change odometer column name from 'odometer' to 'odometer_km'

In [11]:
autos.rename({"odometer":"odometer_km"},axis=1, inplace=True)
autos["odometer_km"].head(5)

0    150000
1    125000
2    125000
3    150000
4     90000
Name: odometer_km, dtype: int64

### Exploring Odometer and Price


In [12]:
autos["odometer_km"].value_counts()

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: odometer_km, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.



In [13]:
autos["price"].unique().shape

(5597,)

In [14]:
autos["price"].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

In [15]:
autos["price"].value_counts().head(20)

0       10778
500      5670
1500     5394
1000     4649
1200     4594
2500     4438
600      3819
3500     3792
800      3784
2000     3432
999      3364
750      3203
650      3150
4500     3053
850      2946
2200     2936
700      2936
1800     2886
900      2874
950      2793
Name: price, dtype: int64

The prices in this column seem rounded, however given there are 2357 unique values in the column, that may just be people's tendency to round prices on the site.

There are 1,421 cars listed with $0 price - given that this is only 2% of the of the cars, we might consider removing these rows. The maximum price is one hundred million dollars, which seems a lot, let's look at the highest prices further.

In [16]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
Name: price, dtype: int64

In [17]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
Name: price, dtype: int64

There are a number of listings with prices below \$30, including about 1,500 at \$0. There are also a small number of listings with very high values, including 14 at around or over $1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. We will keep the \$1 items, but remove anything above \$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [18]:
autos = autos[autos["price"].between(1,1,351000)]
autos["price"].describe()

count    1189.0
mean        1.0
std         0.0
min         1.0
25%         1.0
50%         1.0
75%         1.0
max         1.0
Name: price, dtype: float64

### Exploring the date columns


There are a number of columns with date information:

date_crawled
registration_month
registration_year
ad_created
last_seen
These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings

In [19]:
autos[["date_crawled","ad_created","last_seen"]][0:5]

date_crawled           ad_created            last_seen
60   2016-03-29 15:48:15  2016-03-29 00:00:00  2016-04-06 01:44:38
91   2016-03-28 09:37:01  2016-03-28 00:00:00  2016-04-06 12:15:54
279  2016-03-12 07:26:49  2016-03-11 00:00:00  2016-03-26 23:15:49
422  2016-03-29 18:51:58  2016-03-29 00:00:00  2016-04-06 04:45:50
941  2016-03-06 00:25:25  2016-03-05 00:00:00  2016-04-06 20:19:08

In [20]:
(autos["date_crawled"]
   .str[:10]
   .value_counts(normalize=True,dropna=False)
   .sort_index()

)

2016-03-05    0.024390
2016-03-06    0.015139
2016-03-07    0.040370
2016-03-08    0.039529
2016-03-09    0.028595
2016-03-10    0.038688
2016-03-11    0.030278
2016-03-12    0.023549
2016-03-13    0.014298
2016-03-14    0.031960
2016-03-15    0.035324
2016-03-16    0.029437
2016-03-17    0.029437
2016-03-18    0.015980
2016-03-19    0.037006
2016-03-20    0.042052
2016-03-21    0.042052
2016-03-22    0.045416
2016-03-23    0.034483
2016-03-24    0.033642
2016-03-25    0.037006
2016-03-26    0.029437
2016-03-27    0.033642
2016-03-28    0.037847
2016-03-29    0.033642
2016-03-30    0.021026
2016-03-31    0.034483
2016-04-01    0.027754
2016-04-02    0.025231
2016-04-03    0.026913
2016-04-04    0.034483
2016-04-05    0.019344
2016-04-06    0.004205
2016-04-07    0.003364
Name: date_crawled, dtype: float64

In [21]:
(autos["date_crawled"]
   .str[:10]
   .value_counts(normalize=True,dropna=False)
   .sort_values()

)

2016-04-07    0.003364
2016-04-06    0.004205
2016-03-13    0.014298
2016-03-06    0.015139
2016-03-18    0.015980
2016-04-05    0.019344
2016-03-30    0.021026
2016-03-12    0.023549
2016-03-05    0.024390
2016-04-02    0.025231
2016-04-03    0.026913
2016-04-01    0.027754
2016-03-09    0.028595
2016-03-16    0.029437
2016-03-17    0.029437
2016-03-26    0.029437
2016-03-11    0.030278
2016-03-14    0.031960
2016-03-27    0.033642
2016-03-29    0.033642
2016-03-24    0.033642
2016-03-31    0.034483
2016-03-23    0.034483
2016-04-04    0.034483
2016-03-15    0.035324
2016-03-19    0.037006
2016-03-25    0.037006
2016-03-28    0.037847
2016-03-10    0.038688
2016-03-08    0.039529
2016-03-07    0.040370
2016-03-21    0.042052
2016-03-20    0.042052
2016-03-22    0.045416
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.




In [22]:
(autos["last_seen"]
   .str[:10]
   .value_counts(normalize=True,dropna=False)
   .sort_values()

)

2016-03-06    0.002523
2016-03-18    0.007569
2016-03-07    0.008410
2016-03-13    0.009251
2016-03-09    0.011775
2016-04-01    0.011775
2016-03-23    0.012616
2016-03-19    0.012616
2016-03-11    0.012616
2016-03-08    0.013457
2016-03-21    0.014298
2016-03-10    0.014298
2016-03-22    0.014298
2016-03-26    0.014298
2016-03-14    0.014298
2016-03-16    0.014298
2016-04-03    0.014298
2016-03-15    0.015980
2016-03-24    0.018503
2016-03-31    0.018503
2016-03-29    0.019344
2016-03-28    0.021026
2016-03-20    0.021867
2016-03-17    0.021867
2016-03-30    0.022708
2016-04-02    0.024390
2016-04-04    0.025231
2016-03-12    0.025231
2016-03-25    0.026072
2016-03-27    0.026913
2016-04-07    0.133726
2016-04-05    0.137931
2016-04-06    0.238015
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [24]:
autos["ad_created"].str[:10].unique().shape

(35,)

In [25]:
(autos["ad_created"]
   .str[:10]
   .value_counts(normalize=True,dropna=False)
   .sort_values()

)

2016-03-04    0.001682
2016-04-07    0.003364
2016-04-06    0.004205
2016-03-13    0.014298
2016-03-18    0.015980
2016-03-06    0.015980
2016-04-05    0.018503
2016-03-30    0.021026
2016-04-02    0.023549
2016-03-12    0.024390
2016-03-05    0.025231
2016-04-03    0.026072
2016-03-16    0.029437
2016-03-26    0.029437
2016-03-09    0.029437
2016-04-01    0.030278
2016-03-11    0.030278
2016-03-17    0.031119
2016-03-27    0.031119
2016-03-23    0.031960
2016-03-14    0.031960
2016-03-29    0.033642
2016-03-24    0.033642
2016-03-15    0.033642
2016-03-31    0.034483
2016-04-04    0.035324
2016-03-19    0.037006
2016-03-07    0.037006
2016-03-28    0.037847
2016-03-10    0.038688
2016-03-08    0.039529
2016-03-25    0.039529
2016-03-21    0.040370
2016-03-20    0.042893
2016-03-22    0.047098
Name: ad_created, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.




In [26]:
autos["registration_year"].describe()

count    1189.000000
mean     2008.621531
std       241.713802
min      1000.000000
25%      1996.000000
50%      2000.000000
75%      2005.000000
max      8000.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

